source/particle/interpolator/harmonic_average.cc

```diff
@@ -0,0 +1,150 @@
+
+                for (unsigned int i = 0; i < particle_properties.size(); ++i)
+                  if (selected_properties[i])
+                    cell_properties[i] += 1/particle_properties[i];
+              }
+
+            for (unsigned int i = 0; i < n_particle_properties; ++i)
+              if (selected_properties[i])
+                {
+                  cell_properties[i] = n_particles/cell_properties[i];
+                }
+          }
+        // If there are no particles in this cell use the average of the
+        // neighboring cells.
+        else
+          {
+            std::vector<typename parallel::distributed::Triangulation<dim>::active_cell_iterator> neighbors;
+            GridTools::get_active_neighbors<parallel::distributed::Triangulation<dim> >(found_cell,neighbors);
+
+            unsigned int non_empty_neighbors = 0;
+            for (unsigned int i=0; i<neighbors.size(); ++i)
+              {
+                // Only recursively call this function if the neighbor cell contains
+                // particles (else we end up in an endless recursion)
+                if (particle_handler.n_particles_in_cell(neighbors[i]) == 0)
+                  continue;
+
+                const std::vector<double> neighbor_properties = properties_at_points(particle_handler,
+                                                                                     std::vector<Point<dim> > (1,neighbors[i]->center(true,false)),
+                                                                                     selected_properties,
+                                                                                     neighbors[i])[0];
+
+                for (unsigned int i = 0; i < n_particle_properties; ++i)
+                  if (selected_properties[i])
+                    cell_properties[i] += 1/neighbor_properties[i];
+
+                ++non_empty_neighbors;
+              }
+
+            AssertThrow(non_empty_neighbors != 0,
+                        ExcMessage("A cell and all of its neighbors do not contain any particles. "
+                                   "The `harmonic average' interpolation scheme does not support this case."));
+
+            for (unsigned int i = 0; i < n_particle_properties; ++i)
+              if (selected_properties[i])
+                cell_properties[i] = non_empty_neighbors/cell_properties[i];
+          }
+
+        return std::vector<std::vector<double> > (positions.size(),cell_properties);
+      }
+    }
+  }
+}
+
+
+// explicit instantiations
+namespace aspect
+{
+  namespace Particle
+  {
+    namespace Interpolator
+    {
+      ASPECT_REGISTER_PARTICLE_INTERPOLATOR(HarmonicAverage,
+                                            "harmonic average",
+                                            "Return the harmonic average of all particle properties in the "
+                                            "given cell.")
```

Comment:
Please add something like the following: `If the cell contains no particles, return the harmonic average of the properties in the neighboring cells.`
This is just to make clear that there are two differences compared to `cell_average`, the averaging of particles, and the averaging in case the cell is empty.